contrib/ros/rikai/contrib/ros/convert.py

```diff
@@ -0,0 +1,348 @@
+
+        return False
+
+    def convert(self, message_type: str, value: genpy.message.Message):
+        if message_type in self.CONVERT_MAP:
+            try:
+                return self.CONVERT_MAP[message_type](value)
+            except TypeError as type_err:
+                logging.error(
+                    "Failed to convert type=%s, value=%s, converter=%s: %s",
+                    message_type,
+                    value,
+                    self.CONVERT_MAP[message_type],
+                    type_err,
+                )
+                raise
+
+        array_type_and_size = parse_array(message_type)
+        if array_type_and_size:
+            message_type = array_type_and_size[0]
+            return [self.convert(message_type, v) for v in value]
+
+        return None
+
+    def get_value(self, value, attr_name):
+        return getattr(value, attr_name)
+
+    def array_type(self, value):
+        if value is None:
+            return []
+
+        return list(value)
+
+    def build_record(self, fields):
+        return fields
+
+
+class RikaiConverter(Converter):
+    pass
+
+
+class SparkSchemaConverter(Converter):
+    CONVERT_MAP = {
+        "bool": BooleanType,
+        "char": ByteType,
+        "int8": ByteType,
+        "int16": ShortType,
+        "int32": IntegerType,
+        "int64": LongType,
+        # Converting the unsigned numbers might overflow,
+        # Use a large data instead.
+        "uint8": ShortType,
+        "uint16": IntegerType,
+        "uint32": LongType,
+        "uint64": LongType,
+        "float32": FloatType,
+        "float64": DoubleType,
+        "string": StringType,
+        "time": TimestampType,
+        "duration": LongType,
+        "byte": ByteType,
+        "byte[]": BinaryType,
+        "uint8[]": BinaryType,
+    }
+
+    SCHEMA_CACHE = {}
+
+    def is_supported(
+        self,
+        message_type: str,
+    ) -> bool:
+        if message_type in self.CONVERT_MAP:
+            return True
+        if message_type == "uint64":
+            raise ValueError("uint64 is overflow in Spark")
+
+        # TODO: use local assignment after python 3.8
+        array_type_and_size = parse_array(message_type)
+        if array_type_and_size:
+            return self.is_supported(array_type_and_size[0])
+
+        return False
+
+    def convert(self, message_type: str, value: genpy.message.Message):
+        if message_type in self.CONVERT_MAP:
+            return self.CONVERT_MAP[message_type]()
+        array_type_and_size = parse_array(message_type)
+        if array_type_and_size and self.is_supported(array_type_and_size[0]):
+            return ArrayType(self.convert(array_type_and_size[0], value))
+        return None
+
+    def get_value(self, value, attr_name):
+        msg_type = value._slot_types[value.__slots__.index(attr_name)]
+        if msg_type in self.SCHEMA_CACHE:
+            return self.SCHEMA_CACHE[msg_type]
+        if self.is_supported(msg_type):
```

Comment:
since it’s all early return it doesn’t really matter right now but an `elif` cascade reads a bit cleaner. And also, it’s a bit safe if we accident mess up the early return it’ll return None rather than fall into another branch.